SwiftCompilerSources/Sources/Optimizer/Utilities/BorrowUtils.swift

```diff
@@ -0,0 +1,780 @@
+      case let .beginBorrow(bbi):
+        // Gather the outer enclosing borrow scope.
+        BorrowIntroducers.gather(for: bbi.operand.value, in: &enclosingValues,
+                                 &cache, context)
+      case .loadBorrow, .beginApply, .functionArgument:
+        // There is no enclosing value on this path.
+        break
+      case let .reborrow(reborrow):
+        gather(forReborrow: reborrow, in: &enclosingValues, &cache)
+      }
+    } else {
+      // Handle forwarded guaranteed values.
+      BorrowIntroducers.gather(for: value, in: &enclosingValues,
+                               &cache, context)
+    }
+  }
+  
+  // Given a reborrow, find the enclosing values. Each enclosing value
+  // is represented by one of the following cases, which refer to the
+  // example below:
+  //
+  // dominating owned value -> %value encloses %reborrow_1
+  // owned outer-adjacent phi -> %phi_3 encloses %reborrow_3
+  // dominating outer borrow introducer -> %outerBorrowB encloses %reborrow
+  // outer-adjacent reborrow -> %outerReborrow encloses %reborrow
+  //
+  // Recurse along the use-def phi web until a begin_borrow is
+  // reached. Then find all introducers of the begin_borrow's
+  // operand. At each level, find the outer adjacent phi, if one
+  // exists, otherwise return the most recently found dominating
+  // definition.
+  //
+  // If `reborrow` was already encountered because of a phi cycle,
+  // then no enclosingDefs are added.
+  //
+  // Example:
+  //
+  //         %value = ...
+  //         %borrow = begin_borrow %value
+  //         br one(%borrow)
+  //     one(%reborrow_1 : @guaranteed)
+  //         br two(%value, %reborrow_1)
+  //     two(%phi_2 : @owned, %reborrow_2 : @guaranteed)
+  //         br three(%value, %reborrow_2)
+  //     three(%phi_3 : @owned, %reborrow_3 : @guaranteed)
+  //         end_borrow %reborrow_3
+  //         destroy_value %phi_3
+  //
+  // gather(forReborrow: %reborrow_3) finds %phi_3 by computing
+  // enclosing defs in this order
+  //     (inner -> outer):
+  //
+  //     %reborrow_1 -> %value
+  //     %reborrow_2 -> %phi_2
+  //     %reborrow_3 -> %phi_3
+  //
+  // Example:
+  //
+  //         %outerBorrowA = begin_borrow
+  //         %outerBorrowB = begin_borrow
+  //         %struct = struct (%outerBorrowA, outerBorrowB)
+  //         %borrow = begin_borrow %struct
+  //         br one(%outerBorrowA, %borrow)
+  //     one(%outerReborrow : @guaranteed, %reborrow : @guaranteed)
+  //
+  // gather(forReborrow: %reborrow) finds (%outerReborrow, %outerBorrowB).
+  //
+  private mutating func gather(forReborrow reborrow: Phi,
+                               in enclosingValues: inout Stack<Value>,
+                               _ cache: inout BorrowIntroducers.Cache) {
+
+    guard visitedReborrows.insert(reborrow.value) else {
+      return
+    }
+    // avoid duplicates in the enclosingValues set.
+    var pushedEnclosingValues = ValueSet(context)
```

Comment:
You cannot use `ValueSet` in a recursive function. It will run out of bit positions (for the set) pretty quickly.
Can you define `pushedEnclosingValues` as member property of `EnclosingValues`?

Reply:
Thanks, i meant to convert that ValueSet into a cache. That's done now.